zeppelin-plugins/notebookrepo/mongo/src/main/java/org/apache/zeppelin/notebook/repo/MongoNotebookRepo.java

```diff
@@ -0,0 +1,390 @@
+import org.apache.zeppelin.notebook.NoteInfo;
+import org.apache.zeppelin.user.AuthenticationInfo;
+
+/**
+ * Backend for storing Notebook on MongoDB.
+ */
+public class MongoNotebookRepo implements NotebookRepo {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MongoNotebookRepo.class);
+
+  private ZeppelinConfiguration conf;
+
+  private MongoClient client;
+
+  private MongoDatabase db;
+
+  private MongoCollection<Document> notes;
+
+  private MongoCollection<Document> folders;
+
+  private String folderName;
+
+  public MongoNotebookRepo() {
+  }
+
+  @Override
+  public void init(ZeppelinConfiguration zConf) throws IOException {
+    this.conf = zConf;
+    client = new MongoClient(new MongoClientURI(conf.getMongoUri()));
+    db = client.getDatabase(conf.getMongoDatabase());
+    notes = db.getCollection(conf.getMongoCollection());
+    folderName = conf.getMongoFolder();
+    folders = db.getCollection(folderName);
+
+    if (conf.getMongoAutoimport()) {
+      // import local notes into MongoDB
+      insertFileSystemNotes();
+    }
+  }
+
+  private void insertFileSystemNotes() throws IOException {
+    NotebookRepo vfsRepo = new VFSNotebookRepo();
+    vfsRepo.init(this.conf);
+    Map<String, NoteInfo> infos = vfsRepo.list(null);
+    for (NoteInfo info : infos.values()) {
+      Note note = vfsRepo.get(info.getId(), info.getPath(), null);
+      save(note, null);
+    }
+
+    vfsRepo.close();
+  }
+
+  @Override
+  public Map<String, NoteInfo> list(AuthenticationInfo subject) throws IOException {
+    Map<String, NoteInfo> infos = new HashMap<>();
+
+    Document match = new Document("$match", new Document(Fields.IS_DIR, false));
+    Document graphLookup = new Document("$graphLookup"
+        , new Document("from", folderName)
+        .append("startWith", "$" + Fields.PID)
+        .append("connectFromField", Fields.PID)
+        .append("connectToField", Fields.ID)
+        .append("as", "fullPath"));
+
+
+    ArrayList<Document> list = Lists.newArrayList(match, graphLookup);
+
+    AggregateIterable<Document> aggregate = folders.aggregate(list);
+    for (Document document : aggregate) {
+      String id = document.getString(Fields.ID);
+      String name = document.getString(Fields.NAME);
+      List<Document> fullPath = document.get("fullPath", List.class);
+
+      StringBuilder sb = new StringBuilder();
+      for (Document pathNode : fullPath) {
+        sb.append("/").append(pathNode.getString(Fields.NAME));
+      }
+
+      String fullPathStr = sb.append("/").append(name).toString();
+
+      NoteInfo noteInfo = new NoteInfo(id, fullPathStr);
+      infos.put(id, noteInfo);
+    }
+
+    return infos;
+  }
+
+  @Override
+  public Note get(String noteId, String notePath, AuthenticationInfo subject) throws IOException {
+    ////validate folder exists
+    //String[] pathArray = toPathArray(notePath, false);
+    //findFolder(pathArray);
+    LOG.info("====get{}:{}", noteId, notePath);
```

Comment:
please update - should we keep this or format to remove `====`

Reply:
ok, I'll format it and change it to debug log level.